core/src/main/java/amino/run/kernel/client/KernelClient.java

```diff
@@ -154,4 +299,86 @@ public void copyObjectToServer(InetSocketAddress host, KernelOID oid, KernelObje
                     MicroServiceReplicaNotFoundException {
         getServer(host).copyKernelObject(oid, object);
     }
+
+    /**
+     * Method to measure the kernel server metrics with respect to other remote servers and store
+     * them in {@link amino.run.kernel.client.KernelClient#metrics}. These collected metrics are
+     * reported to OMS in heartBeats
+     */
+    private void measureMetrics(final KernelServerInfo serverInfo) {
+        KernelServer server = serverInfo.remoteRef;
+        int dataLength = serverInfo.data.len;
+        NodeMetric metric = serverInfo.metric;
+        int period = serverInfo.metricPollPeriod;
+        serverInfo.refCount++;
+        if (serverInfo.refCount == Integer.MAX_VALUE) {
+            /* Reset to 1 when it is about to overflow */
+            serverInfo.refCountOnHeartBeat = serverInfo.refCount = 1;
+        }
+
+        try {
+            /* Forward time = Request Serialization + Network Delay + Deserialization + Object/call Dispatch time
+            + processing time + Optionally connection establishment time */
+            /* Return time = Response Serialization + Network Delay + Deserialization + Get current nano time at end */
+            /* Forward time would be comparatively higher than return time. But to avoid the clock sync issues
+            between client and server, took the round trip time(RTT) as latency
+             */
+
+            /* Make an empty heartbeat to ensure session is established and cached before measurement */
+            server.receiveHeartBeat();
+
+            /* Measure latency */
+            long t1 = System.nanoTime();
+            server.receiveHeartBeat();
+            long t2 = System.nanoTime();
+
+            /* Measure data transfer rate by sending some arbitrary data */
+            server.receiveHeartBeat(serverInfo.data);
+            long t3 = System.nanoTime();
+            if (t3 - t2 < t2 - t1) {
```

Comment:
Please add a comment to describe what you're doing in this block, e.g. "data too small relative to line latency - try again immediately with one more step of data".

Reply:
Added comment